worlds/ufo50/games/vainger/rules.py

```diff
@@ -0,0 +1,248 @@
+from typing import TYPE_CHECKING, Dict
+from BaseClasses import Region, CollectionState
+from worlds.generic.Rules import set_rule, CollectionRule
+
+if TYPE_CHECKING:
+    from ... import UFO50World
+
+# adapted from Barbuta, thanks Scipio! <3
+
+# name upgrades for convenience
+heat_mod = "Vainger - Heat Mod"
+multi_mod = "Vainger - Multi Mod"
+pulse_mod = "Vainger - Pulse Mod"
+force_mod = "Vainger - Force Mod"
+
+stabilizer = "Vainger - Stabilizer"
+shield_upgrade = "Vainger - Shield Upgrade"
+keycode_A = "Vainger - Key Code A"
+keycode_B = "Vainger - Key Code B"
+keycode_C = "Vainger - Key Code C"
+keycode_D = "Vainger - Key Code D"
+security_clearance = "Vainger - Progressive Security Clearance"
+
+# can the player do a given hell run?
+# TODO: option for this
+def hell_run(shield_upgrades_required: int, is_vanilla: bool, state: CollectionState, world: "UFO50World") -> bool:
+    return state.count(shield_upgrade, world.player) >= shield_upgrades_required
+
+# can the player beat a given boss? currently checks how many mods the player has,
+# whether they have a stabilizer, and if they hit a vibes-based shield threshold.
+# TODO: option for this
+# difficulty must be between 0 and 4 inclusive
+def boss_logic(difficulty: int, state: CollectionState, world:"UFO50World") -> bool:
+    player = world.player
+    if not state.has_from_list_unique([heat_mod, multi_mod, pulse_mod, force_mod], player, difficulty):
+        return False
+    if (difficulty == 4) and not state.has(stabilizer, player):
+        return False
+    shield_upgrades_required = [0, 0, 5, 10, 15][difficulty]
+    return (state.count(shield_upgrade, player) >= shield_upgrades_required)
```

Comment:
```suggestion
    if difficulty == 4 and not state.has(stabilizer, player):
        return False
    shield_upgrades_required = [0, 0, 5, 10, 15][difficulty]
    return state.has(shield_upgrade, player, shield_upgrades_required)
```
`state.has` can take a quantity

The () are not required around `difficulty == 4` (though you can keep it if you want)